sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -620,6 +620,12 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
     // We can return what the children return. Same thing should happen in the codegen path.
     if (DataType.equalsStructurally(from, to)) {
       identity
+    } else if (from == NullType) {
+      // According to `canCast`, NullType can be casted to any type.
+      // For primitive types, we don't reach here because the guard of `nullSafeEval`.
+      // But for nested types like struct, we might reach here for nested null type field.
+      // We won't call the returned function actually, but returns a placeholder.
+      _ => throw new SparkException(s"should not directly cast from NullType to $to.")
```

Comment:
when will we hit this exception?

Reply:
In the case of casting a nested field in a struct or array, if we find it is a null, we don't call the cast function but set the destination as null directly. If we call this cast function, we hit this exception.

Reply:
We do the thing as follow:
```scala
buildCast[InternalRow](_, row => {
  val newRow = new GenericInternalRow(from.fields.length)
  var i = 0
  while (i < row.numFields) {
    newRow.update(i,
      // We don't call cast function but directly set a null, if finding a null.
      if (row.isNullAt(i)) null else castFuncs(i)(row.get(i, from.apply(i).dataType)))
    i += 1
  }
  newRow
})
```

Reply:
Which means, it's a bug if the exception is thrown?

Reply:
If reaching here, means we call the cast function for a null, unnecessarily.